README.md

# flutter_responsive_scale
**A flutter plugin for adapting screen and font size.Let your UI display a reasonable layout on different screen sizes by using device scale!**

## Demo
![ezgif-4-0040112d85](https://user-images.githubusercontent.com/70679949/214622291-47580841-f4ef-45f0-bfec-4e42f51b8625.gif)

https://user-images.githubusercontent.com/70679949/214531030-25c9f4d9-9ca5-4067-9a2b-9071cd6de115.mp4

## Features

- Responsive UI
- Simple UI Utils
- Easy to use
- Responsive Query option - fontScale, yScale and width scale

## Usage

### Add dependency

Please check the latest version before installation.
If there is any problem with the new version, please use the previous version

```yaml
dependencies:
  flutter:
    sdk: flutter
  # add flutter_responsive_scale
  flutter_responsive_scale: ^{latest version}
```

### Add the following imports to your Dart code

```dart
import 'package:flutter_responsive_scale/flutter_responsive_scale.dart';
```

### Property

| Property              | Type     | Parameter                                                       | Description                                                                                                                                                                                                                    |
|-----------------------|----------|-----------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| context.Scale(size)   | Function | double required                                                 | Pixels scaled per device from design. Where One pixel on a 160px screen equals two pixels on a 320px screen. Also and alias for scaleX.                                                                                        |
| context.fontScale(14) | Function | double required                                                 | Relative to the font-size setting of the actual device.                                                                                                                                                                        |
| context.scaleY()      | Function | double required                                                 | Reference device configuration -  Reference width of device. Defaults to 414px (iPhone XS Max)  Reference height of device. Defaults to 896px (iPhone XS Max),  Should allow automatic scaling of font sizes. Defaults to true |
| ResponsiveScaleConfig | Clase    | height (double), width (double), allowFontScaling (bool - true) | Pixels scaled per device from design in the Vertical direction. Where One pixel on a 160px screen equals two pixels on a 320px screen.                                                                                         |

### Initialize and set the fit size and font size to scale according to the system's "font size" accessibility option

Please set the scale config of the design draft before use, the width, height of the design draft and allowFontScaling.

```dart

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return ResponsiveScale(
      config: const ResponsiveScaleConfig(),
      child: MaterialApp(
        title: 'Responsive scale demo',
        theme: ThemeData.dark(),
        home: const ScaleTestPage(
          title: 'Hi , Flutter dev - ',
        ),
      ),
    );
  }
}

```

## Examples

```dart
import 'package:flutter/material.dart';
import 'package:flutter_responsive_scale/flutter_responsive_scale.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return ResponsiveScale(
      config: const ResponsiveScaleConfig(),
      child: MaterialApp(
        title: 'Responsive scale demo',
        theme: ThemeData.dark(),
        home: const ScaleTestPage(
          title: 'Hi , Flutter dev - ',
        ),
      ),
    );
  }
}

class ScaleTestPage extends StatelessWidget {
  const ScaleTestPage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(title)),
      body: SingleChildScrollView(
        padding: EdgeInsets.symmetric(
          horizontal: context.scale(7),
        ),
        child: Center(
          child: Column(
            children: <Widget>[
              SizedBox(height: context.scale(16)),
              Text(
                'Designed at ${ResponsiveScale.of(context).width} x ${ResponsiveScale.of(context).height}',
                style: TextStyle(
                  fontSize: context.fontScale(16),
                ),
                textAlign: TextAlign.center,
              ),
              SizedBox(
                height: context.scale(16),
              ),
              Container(
                alignment: Alignment.topLeft,
                padding: EdgeInsets.only(
                  bottom: context.scaleY(10),
                ),
                child: Text(
                  'With scale box',
                  style: TextStyle(
                    fontSize: context.fontScale(14),
                  ),
                  textAlign: TextAlign.left,
                ),
              ),
              RenderBoxes(
                size: Size.square(
                  context.scale(100),
                ),
                color: Colors.red,
                label: const ["100dp", "100dp"],
              ),
              SizedBox(
                height: context.scale(16),
              ),
              Container(
                alignment: Alignment.topLeft,
                padding: EdgeInsets.only(
                  bottom: context.scaleY(10),
                ),
                child: Text(
                  'Without scale box',
                  style: TextStyle(
                    fontSize: context.fontScale(14),
                  ),
                  textAlign: TextAlign.left,
                ),
              ),
              const RenderBoxes(
                size: Size.square(
                  100,
                ),
                color: Colors.blue,
                label: ["100", "100"],
              ),
              SizedBox(
                height: context.scale(16),
              ),
              const Text(
                'The text font size is fixed at 24',
                style: TextStyle(
                  fontSize: 24,
                ),
                textAlign: TextAlign.center,
              ),
              SizedBox(
                height: context.scale(16),
              ),
              Text(
                'The text font size is flexible at 24',
                style: TextStyle(
                  fontSize: context.fontScale(24),
                ),
                textAlign: TextAlign.center,
              ),
              SizedBox(
                height: context.scale(48),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class RenderBoxes extends StatelessWidget {
  const RenderBoxes({required this.size, this.color, required this.label, Key? key}) : super(key: key);

  final Size size;
  final Color? color;
  final List<String> label;

  @override
  Widget build(BuildContext context) {
    return Row(
      children: List.generate(
        4,
            (_) => Container(
          width: size.width,
          height: size.height,
          color: color,
          child: Center(
            child: Text(
              label.join("\nx\n"),
              style: TextStyle(
                fontSize: context.fontScale(14),
              ),
              textAlign: TextAlign.center,
            ),
          ),
        ),
      ),
    );
  }
}

```
## Motivation
Main motivation behind this package was one of my project feature was required a responsive UI and simple device query for different platform.
## Authors
- [@subhashDev11](https://github.com/subhashDev11)

## 🚀 About Me
I'm a flutter developer...

## 🔗 Links
[![portfolio](https://img.shields.io/badge/my_portfolio-000?style=for-the-badge&logo=ko-fi&logoColor=white)](https://subhashdev121.github.io/subhash/#/)
[![linkedin](https://img.shields.io/badge/linkedin-0A66C2?style=for-the-badge&logo=linkedin&logoColor=white)](https://www.linkedin.com/in/subhashcs)
[![medium](https://img.shields.io/badge/medium-000?style=for-the-badge&logo=medium&logoColor=white)](https://medium.com/@subhashchandrashukla)

![Logo](https://i.ibb.co/2szbbHF/code-xposer.png)

## License

[MIT](https://choosealicense.com/licenses/mit/)


CHANGELOG.md
## 0.0.1
* flutter_responsive_scale initial version.

## 0.0.2
* Example image added.



lib/src/extensions.dart
/*
* About author
* Name - SUBHASH CHANDRA SHUKLA
* Email - subhashchandras7318@gmail.com
* Contact Number - 7318459902
* Github - https://github.com/subhashDev11
* Linked In - https://www.linkedin.com/in/subhashcs/
* */

import 'package:flutter/widgets.dart';
import 'package:flutter_responsive_scale/src/device_scale.dart';

import 'responsive_scale.dart';

extension BuildContextExtensions on BuildContext {
  DeviceScale get _scale {
    final data = MediaQuery.of(this);
    return DeviceScale(
      size: data.size,
      textScaleFactor: data.textScaleFactor,
      config: ResponsiveScale.of(this),
    );
  }

  /// Pixels scaled per device from design. Where One pixel on a 160px screen equals two pixels on a 320px screen.
  /// Also and alias for scaleX
  double scale(num width) => _scale.scale(width);

  /// Pixels scaled per device from design in the Vertical direction. Where One pixel on a 160px screen equals two pixels on a 320px screen.
  double scaleY(num height) => _scale.scaleY(height);

  /// Relative to the font-size setting of the actual device
  double fontScale(num fontSize) => _scale.fontScale(fontSize);
}


lib/src/device_scale.dart
/*
* About author
* Name - SUBHASH CHANDRA SHUKLA
* Email - subhashchandras7318@gmail.com
* Contact Number - 7318459902
* Github - https://github.com/subhashDev11
* Linked In - https://www.linkedin.com/in/subhashcs/
* */

import 'dart:math';

import 'package:flutter/widgets.dart';

import 'responsive_scale_config.dart';

class DeviceScale {
  factory DeviceScale({
    required Size size,
    required double textScaleFactor,
    required ResponsiveScaleConfig config,
  }) {
    if (_instance != null &&
        size == _instance!.size &&
        textScaleFactor == _instance!.textScaleFactor &&
        config == _instance!.config) {
      return _instance!;
    }
    return _instance = DeviceScale._(
        size: size, textScaleFactor: textScaleFactor, config: config);
  }

  DeviceScale._({
    required this.size,
    required this.textScaleFactor,
    required this.config,
  });

  @visibleForTesting
  static void reset() => _instance = null;

  static DeviceScale? _instance;

  /// Scale configuration of reference device
  final ResponsiveScaleConfig config;

  /// Physical size of current device
  final Size size;

  /// Scale factor of font size on current device
  final double textScaleFactor;

  double get _widthScale => size.width / config.width;

  /// Pixels scaled per device from design. Where One pixel on a 160px screen equals two pixels on a 320px screen.
  /// Also and alias for scaleX
  double scale(num dimension) => dimension * _widthScale;

  double get _heightScale => size.height / config.height;

  /// Pixels scaled per device from design. Where One pixel on a 160px screen equals two pixels on a 320px screen.
  double scaleY(num dimension) => dimension * _heightScale;

  /// Relative to the font-size setting of the current device
  double fontScale(num fontSize) => config.allowFontScaling
      ? (fontSize * min(_widthScale, _heightScale)) / textScaleFactor
      : fontSize.toDouble();

  @override
  String toString() => '$runtimeType('
      'textScaleFactor: $textScaleFactor, '
      'width: ${size.width}, '
      'height: ${size.height}'
      ')';
}


lib/src/responsive_scale.dart
/*
* About author
* Name - SUBHASH CHANDRA SHUKLA
* Email - subhashchandras7318@gmail.com
* Contact Number - 7318459902
* Github - https://github.com/subhashDev11
* Linked In - https://www.linkedin.com/in/subhashcs/
* */

import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_responsive_scale/flutter_responsive_scale.dart';

class ResponsiveScale extends InheritedWidget {
  const ResponsiveScale({
    Key? key,
    this.config = const ResponsiveScaleConfig(),
    required Widget child,
  }) : super(key: key, child: child);

  final ResponsiveScaleConfig config;

  static ResponsiveScaleConfig of(BuildContext context) {
    final scaleAware =
        context.dependOnInheritedWidgetOfExactType<ResponsiveScale>();
    if (scaleAware != null) {
      return scaleAware.config;
    }
    throw FlutterError.fromParts(<DiagnosticsNode>[
      ErrorSummary(
          'ResponsiveScale.of() called with a context that does not contain a ResponsiveScale.'),
      ErrorDescription(
          'No ResponsiveScale ancestor could be found starting from the context that was passed '
          'to ResponsiveScale.of(). This can happen because you do not have a ResponsiveScale widget '
          'or it can happen if the context you use comes from a widget above it.'),
      context.describeElement('The context used was')
    ]);
  }

  @override
  bool updateShouldNotify(ResponsiveScale oldWidget) =>
      config != oldWidget.config;

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties
        .add(DiagnosticsProperty<ResponsiveScaleConfig>('config', config));
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_responsive_scale/flutter_responsive_scale.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return ResponsiveScale(
      config: const ResponsiveScaleConfig(),
      child: MaterialApp(
        title: 'Responsive scale demo',
        theme: ThemeData.dark(),
        home: const ScaleTestPage(
          title: 'Hi , Flutter dev - ',
        ),
      ),
    );
  }
}

class ScaleTestPage extends StatelessWidget {
  const ScaleTestPage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(title)),
      body: SingleChildScrollView(
        padding: EdgeInsets.symmetric(
          horizontal: context.scale(7),
        ),
        child: Center(
          child: Column(
            children: <Widget>[
              SizedBox(height: context.scale(16)),
              Text(
                'Designed at ${ResponsiveScale.of(context).width} x ${ResponsiveScale.of(context).height}',
                style: TextStyle(
                  fontSize: context.fontScale(16),
                ),
                textAlign: TextAlign.center,
              ),
              SizedBox(
                height: context.scale(16),
              ),
              Container(
                alignment: Alignment.topLeft,
                padding: EdgeInsets.only(
                  bottom: context.scaleY(10),
                ),
                child: Text(
                  'With scale box',
                  style: TextStyle(
                    fontSize: context.fontScale(14),
                  ),
                  textAlign: TextAlign.left,
                ),
              ),
              RenderBoxes(
                size: Size.square(
                  context.scale(100),
                ),
                color: Colors.red,
                label: const ["100dp", "100dp"],
              ),
              SizedBox(
                height: context.scale(16),
              ),
              Container(
                alignment: Alignment.topLeft,
                padding: EdgeInsets.only(
                  bottom: context.scaleY(10),
                ),
                child: Text(
                  'Without scale box',
                  style: TextStyle(
                    fontSize: context.fontScale(14),
                  ),
                  textAlign: TextAlign.left,
                ),
              ),
              const RenderBoxes(
                size: Size.square(
                  100,
                ),
                color: Colors.blue,
                label: ["100", "100"],
              ),
              SizedBox(
                height: context.scale(16),
              ),
              const Text(
                'The text font size is fixed at 24',
                style: TextStyle(
                  fontSize: 24,
                ),
                textAlign: TextAlign.center,
              ),
              SizedBox(
                height: context.scale(16),
              ),
              Text(
                'The text font size is flexible at 24',
                style: TextStyle(
                  fontSize: context.fontScale(24),
                ),
                textAlign: TextAlign.center,
              ),
              SizedBox(
                height: context.scale(48),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class RenderBoxes extends StatelessWidget {
  const RenderBoxes({required this.size, this.color, required this.label, Key? key}) : super(key: key);

  final Size size;
  final Color? color;
  final List<String> label;

  @override
  Widget build(BuildContext context) {
    return Row(
      children: List.generate(
        4,
        (_) => Container(
          width: size.width,
          height: size.height,
          color: color,
          child: Center(
            child: Text(
              label.join("\nx\n"),
              style: TextStyle(
                fontSize: context.fontScale(14),
              ),
              textAlign: TextAlign.center,
            ),
          ),
        ),
      ),
    );
  }
}
